### PR TITLE
better env support

### PR DESCRIPTION
**Pull request by drusellers**
to support Staging vs Test vs Prod

### CODE REVIEW

File: src/Chassis/Env/ContainerBuilderExtensions.cs [hunk 1]
```diff
@@ -7,19 +8,29 @@ namespace Chassis.Env
...
             {
-                if (AppEnv.ENV == AppEnv.TEST) return cxt.Resolve<TTestType>();
-                return cxt.Resolve<TProdType>();
-            }).As<TService>();
+                e.On<TProdType>(AppEnv.PRODUCTION);
+                e.On<TTestType>(AppEnv.TEST);
+            });
+        }
+
+        public static void RegisterEnvironmentType<TService>(this ContainerBuilder builder, Action<EnvSwitcher<TService>> action)
```
Comment (BrianHarboldt) [inline src/Chassis/Env/ContainerBuilderExtensions.cs:23]: :clap:

File: src/Chassis/Env/EnvSwitcher.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            foreach (var type in _types)
+            {
+                builder.RegisterType(type).AsSelf();
+            }
+        }
+
+        internal TService Resolve(IComponentContext cxt)
+        {
+            var t = _maps[AppEnv.ENV];
```
Comment (BrianHarboldt) [inline src/Chassis/Env/EnvSwitcher.cs]: Should we do a safe lookup or do we want the exception instead?

Comment (drusellers) [inline src/Chassis/Env/EnvSwitcher.cs]: let me test what happens if I do a `null`

Comment (drusellers) [inline src/Chassis/Env/EnvSwitcher.cs]:
```
Autofac.Core.DependencyResolutionException : An exception was thrown while executing a resolve operation. See the InnerException for details. ---> The container does not have an environment registration for BAD for service type IService (See inner exception for details.)
  ----> System.Exception : The container does not have an environment registration for BAD for service type IService
```

Added better error message

File: src/Chassis/Env/EnvSwitcher.cs [hunk 3]
```diff
@@ -27,8 +27,13 @@ internal void RegisterTypes(ContainerBuilder builder)
 
         internal TService Resolve(IComponentContext cxt)
         {
-            var t = _maps[AppEnv.ENV];
-            return cxt.Resolve(t) as TService;
+            if (_maps.ContainsKey(AppEnv.ENV))
```
Comment (BrianHarboldt) [inline src/Chassis/Env/EnvSwitcher.cs:33]: `AppEnv.ENV != null &&`
Or do we want the standard puke for looking up a dictionary by `null`?

Comment (drusellers) [inline src/Chassis/Env/EnvSwitcher.cs:33]: you won't ever get `null` 

https://github.com/mag-dev/chassis/blob/master/src/tests/Chassis.UnitTests/Env/AppEnvTests.cs#L46

**Comment by BrianHarboldt** [comment]
:shipit:

File: src/Chassis/Env/EnvSwitcher.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        where TService : class
+    {
+        readonly List<Type> _types = new List<Type>();
+        readonly Dictionary<AppEnv, Type> _maps = new Dictionary<AppEnv, Type>();
+
+        public void On<TType>(AppEnv env)
+            where TType : class
+        {
+            _types.Add(typeof(TType));
+            _maps.Add(env, typeof(TType));
```
Comment (sebastianhoff) [inline src/Chassis/Env/EnvSwitcher.cs:20]: Do you want to check before adding if `env` already exists in `_maps` or just let the dictionary exception throw?

File: src/Chassis/Env/EnvSwitcher.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        internal void RegisterTypes(ContainerBuilder builder)
+        {
+            foreach (var type in _types)
```
Comment (sebastianhoff) [inline src/Chassis/Env/EnvSwitcher.cs]: Could use `_maps.Values` and remove `_types`.

Comment (drusellers) [inline src/Chassis/Env/EnvSwitcher.cs]: :clap:

File: src/Chassis/Env/EnvSwitcher.cs (hunk 4 above)
Comment (drusellers) [inline src/Chassis/Env/EnvSwitcher.cs:20]: Good call, added an exception.

**Pull request merged**